File: src/carrot_switch/store/skill.py
"""Skill metadata local store."""
from pathlib import Path
from carrot_switch.store import STORE_ROOT, _ensure_dir, _now_iso
from carrot_switch.config import read_jsonc, write_jsonc


def _store_path(agent: str) -> Path:
    return STORE_ROOT / "skills" / f"{agent}.json"


def load(agent: str) -> dict:
    """Load local skill store. Returns {"skills": {...}}."""
    path = _store_path(agent)
    if not path.exists():
        return {"skills": {}}
    return read_jsonc(path)


def save(agent: str, data: dict) -> None:
    """Persist skill store to disk."""
    _ensure_dir(_store_path(agent))
    write_jsonc(_store_path(agent), data)


def record_install(agent: str, name: str, source: str, source_type: str, path: str) -> None:
    """Record a skill installation."""
    data = load(agent)
    data["skills"][name] = {
        "source": source,
        "source_type": source_type,
        "installed_at": _now_iso(),
        "path": path,
    }
    save(agent, data)


def record_uninstall(agent: str, name: str) -> None:
    """Remove a skill record."""
    data = load(agent)
    data["skills"].pop(name, None)
    save(agent, data)


def get_skill_meta(agent: str, name: str) -> dict | None:
    """Get metadata for a specific skill."""
    data = load(agent)
    return data["skills"].get(name)

File: server/src/lib/store/skill.ts
import { join } from "path";
import { homedir } from "os";
import { existsSync, mkdirSync } from "fs";
import { readJsonc, writeJsonc } from "../jsonc.js";

const STORE_ROOT = join(homedir(), "AppData", "Roaming", ".carrotswitch");

function storePath(agent: string): string {
  return join(STORE_ROOT, "skills", `${agent}.json`);
}

function nowIso(): string {
  return new Date().toISOString().replace(/\.\d{3}Z$/, "Z");
}

function ensureDir(filePath: string): void {
  const dir = filePath.substring(0, filePath.lastIndexOf("\\") !== -1 ? filePath.lastIndexOf("\\") : filePath.lastIndexOf("/"));
  if (!existsSync(dir)) {
    mkdirSync(dir, { recursive: true });
  }
}

export function load(agent: string): Record<string, any> {
  const path = storePath(agent);
  if (!existsSync(path)) return { skills: {} };
  return readJsonc(path);
}

export function save(agent: string, data: Record<string, any>): void {
  const path = storePath(agent);
  ensureDir(path);
  writeJsonc(path, data);
}

export function recordInstall(
  agent: string,
  name: string,
  source: string,
  sourceType: string,
  skillPath: string
): void {
  const data = load(agent);
  if (!data.skills) data.skills = {};
  data.skills[name] = {
    source,
    sourceType,
    installedAt: nowIso(),
    path: skillPath,
  };
  save(agent, data);
}

export function recordUninstall(agent: string, name: string): void {
  const data = load(agent);
  if (data.skills) {
    delete data.skills[name];
  }
  save(agent, data);
}

export function getSkillMeta(agent: string, name: string): Record<string, any> | null {
  const data = load(agent);
  return data.skills?.[name] || null;
}
